### PR TITLE
Automatic API documentation and HTTP input/output validation

### PR DESCRIPTION
**Pull request by cadent-dgates**

### CODE REVIEW

**Comment by cadent-dgates** [comment]
@cadent-mrichman 

This is the current list of what needs to be done, in prioritized order.

[ ] Write tests for `porter/schemas/openapi.py` (`tests/test_schemas_openapi.py`)
[ ] Write tests that validate the `PredictionService` class response/request validations are correctly validating the input/output data. Use `mock.patch` on `porter.services.api.request_json` to test how prediction service instances behave when called and `request_json` returns valid/invalide data according to the `feature_schema` parameters passed.

These also need to be done, we'll prioritize after the tests above are implemented.

[ ] Fix remaining tests that broke.
[ ] Address TODOs in code.
[ ] Add new tests where needed.

**Comment by cadent-mrichman** [comment]
@cadent-dgates The latest changes allow `test_get_post_data_prediction_schema()` to pass for all combinations of `batch_prediction` and `validate_request_data`, _provided that_ `validate_response_data` is False. When `validate_response_data` is True, a KeyError turns up in `PredictionService.__call__()`.  I'm not yet sure if this is a problem with that method or with the test.

**Comment by cadent-mrichman** [comment]
Latest commit partially addresses #17.  We may want to systematically upgrade all requirements + extras and confirm the tests still pass.

File: tests/test_services.py [hunk 1]
```diff
@@ -471,6 +378,176 @@ def test_constructor_fail(self):
...
+            batch_prediction=False,
+            feature_schema=feature_schema,
+        )
+        # check _request_schemas
+        self.assertIsInstance(prediction_service._request_schemas['POST'], openapi.Object)
+        # check request_schemas
+        request_schema = prediction_service.request_schemas['POST']
+        self.assertIsInstance(request_schema, openapi.RequestSchema)
+        # check that id field was inserted
+        self.assertIn('id', request_schema.api_obj.properties)
```
Comment (cadent-dgates) [inline tests/test_services.py:410]: @cadent-mrichman also assert that the features are in the object properties.

**Comment by cadent-dgates** [comment]
This also resolves #17

**Comment by cadent-mrichman** [comment]
This looks to me like it's ready to go forward.

**Pull request merged**